File: common/RtpPacket.cpp
#include <winsock2.h>
#include <ws2tcpip.h>

#pragma comment(lib, "ws2_32.lib")
#include "RtpPacket.h"
#include <cstring>

RtpPacket::RtpPacket()
{
    // RTP header fields (example values)
    header[0] = (2 << 6);   // Version 2, no padding, no extension, 0 CSRC
    header[1] = 26;

    seqNum = 0;
    timestamp = 0;
    ssrc = 123456; // pick random SSRC

    offset = 0;
}

void RtpPacket::beginFrame(const uint8_t* frameData, int frameSize)
{
    // Reset payload and offset for new frame
    payload.assign(frameData, frameData + frameSize);
    offset = 0;

    // Update timestamp per frame (example: +3600)
    timestamp += TIMESTAMP_INCREMENT;

    // Prepare RTP header
    header[0] = (2 << 6);  // Version 2
    header[1] = 26;        // Payload type 26 for JPEG (static)
}

bool RtpPacket::getNextPacket(uint8_t* outBuffer, int& outSize)
{
    const int MAX_RTP_PAYLOAD = 1200;   // safe MTU-sized payload

    if (offset >= payload.size())
        return false; // no more packets

    // Compute packet payload size
    int remaining = payload.size() - offset;
    bool isLastPacket = (remaining <= MAX_RTP_PAYLOAD);
    int packetPayload = remaining > MAX_RTP_PAYLOAD ? MAX_RTP_PAYLOAD : remaining;

    // Build RTP header
    // dont need to use htons() and htonl() since we set byte manually in this lab
    uint16_t seq = seqNum++;
    uint32_t ts = timestamp;
    uint32_t id = ssrc;

    // Copy header bytes
    std::memcpy(outBuffer, header, 12);

    if (isLastPacket) {
        outBuffer[1] |= 0x80; // Marker = 1
    }
    else {
        outBuffer[1] &= 0x7F; // Marker = 0
    }

    // use & 0xFF to ensure only the last 8 bits are taken
    // Set sequence number
    outBuffer[2] = (seq >> 8) & 0xFF;
    outBuffer[3] = seq & 0xFF;

    // Timestamp
    outBuffer[4] = (ts >> 24) & 0xFF;
    outBuffer[5] = (ts >> 16) & 0xFF;
    outBuffer[6] = (ts >> 8) & 0xFF;
    outBuffer[7] = ts & 0xFF;

    // SSRC
    outBuffer[8] = (id >> 24) & 0xFF;
    outBuffer[9] = (id >> 16) & 0xFF;
    outBuffer[10] = (id >> 8) & 0xFF;
    outBuffer[11] = id & 0xFF;

    // Copy payload chunk
    std::memcpy(outBuffer + 12, payload.data() + offset, packetPayload);

    // Output size includes header + payload
    outSize = 12 + packetPayload;

    // Advance offset
    offset += packetPayload;

    return true;
}

File: server/rtp_packet.py
from __future__ import annotations

from dataclasses import dataclass
import os
import random
from typing import Optional


MAX_RTP_PAYLOAD = 1200
PAYLOAD_TYPE_JPEG = 26
TIMESTAMP_INCREMENT = 3600


@dataclass
class RtpPacketizer:
    ssrc: int = random.randint(1, 0xFFFFFFFF)
    seq_num: int = 0
    timestamp: int = 0
    _payload: bytes = b""
    _offset: int = 0

    def begin_frame(self, frame_data: bytes) -> None:
        self._payload = frame_data
        self._offset = 0
        self.timestamp = (self.timestamp + TIMESTAMP_INCREMENT) & 0xFFFFFFFF

    def get_next_packet(self) -> Optional[bytes]:
        if self._offset >= len(self._payload):
            return None

        remaining = len(self._payload) - self._offset
        is_last = remaining <= MAX_RTP_PAYLOAD
        payload_len = remaining if is_last else MAX_RTP_PAYLOAD

        header = bytearray(12)
        header[0] = 2 << 6
        header[1] = PAYLOAD_TYPE_JPEG | (0x80 if is_last else 0x00)

        seq = self.seq_num & 0xFFFF
        self.seq_num = (self.seq_num + 1) & 0xFFFF

        header[2] = (seq >> 8) & 0xFF
        header[3] = seq & 0xFF

        ts = self.timestamp
        header[4] = (ts >> 24) & 0xFF
        header[5] = (ts >> 16) & 0xFF
        header[6] = (ts >> 8) & 0xFF
        header[7] = ts & 0xFF

        ssrc = self.ssrc
        header[8] = (ssrc >> 24) & 0xFF
        header[9] = (ssrc >> 16) & 0xFF
        header[10] = (ssrc >> 8) & 0xFF
        header[11] = ssrc & 0xFF

        chunk = self._payload[self._offset:self._offset + payload_len]
        self._offset += payload_len

        return bytes(header) + chunk
